src/test/java/hudson/plugins/git/extensions/impl/CloneOptionHonorRefSpecTest.java

```diff
@@ -0,0 +1,149 @@
+import hudson.plugins.git.GitSCM;
+import hudson.plugins.git.UserRemoteConfig;
+import hudson.tasks.BatchFile;
+import hudson.tasks.Builder;
+import hudson.tasks.Shell;
+import org.jenkinsci.plugins.gitclient.JGitTool;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.jvnet.hudson.test.Issue;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import static org.hamcrest.MatcherAssert.*;
+import static org.hamcrest.Matchers.*;
+
+@RunWith(Parameterized.class)
+public class CloneOptionHonorRefSpecTest extends AbstractGitTestCase {
+
+    private final String refSpecName;
+    private final Boolean honorRefSpec;
+
+    private static final Random random = new Random();
+    private FreeStyleProject project;
+    private String refSpecExpectedValue;
+
+    public CloneOptionHonorRefSpecTest(String refSpecName, Boolean honorRefSpec) {
+        this.refSpecName = refSpecName;
+        this.honorRefSpec = honorRefSpec;
+    }
+
+    @Parameterized.Parameters(name = "{0}-{1}")
+    public static Collection permuteRefSpecVariable() {
+        List<Object[]> values = new ArrayList<>();
+        // Should behave same with honor refspec enabled or disabled
+        boolean honorRefSpec = random.nextBoolean();
+
+        String[] keys = {
+                "JOB_NAME", // Variable set by Jenkins
+                (Functions.isWindows() ? "USERNAME" : "USER"), // Variable set by the operating system
+                "USER_SELECTED_BRANCH_NAME" // Parametrised build param
+        };
+
+        for (String refSpecName : keys) {
+            Object[] combination = {refSpecName, true};
+            values.add(combination);
+            honorRefSpec = !honorRefSpec;
+        }
+
+        return values;
+    }
+
+    private static Builder createEnvEchoBuilder(String envVarName) {
+        if (Functions.isWindows()) {
+            return new BatchFile(String.format("echo %s=%%%s%%", envVarName, envVarName));
+        }
+        return new Shell(String.format("echo \"%s=${%s}\"", envVarName, envVarName));
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+
+        // Setup job beforehand to get expected value of the environment variable
+        project = createFreeStyleProject();
+        project.addProperty(new ParametersDefinitionProperty(
+                new StringParameterDefinition("USER_SELECTED_BRANCH_NAME", "user_branch")
+        ));
+        project.getBuildersList().add(createEnvEchoBuilder(refSpecName));
+
+        final FreeStyleBuild b = project.scheduleBuild2(0).get();
+        rule.assertBuildStatus(Result.SUCCESS, b);
+
+        List<String> logs = b.getLog(50);
+        for (String line : logs) {
+            if (line.startsWith(refSpecName + '=')) {
+                refSpecExpectedValue = line.split("=")[1];
```

Comment:
@MarkEWaite this seems to be causing a test failure in the BOM, perhaps because there is some environment variable with an empty value: https://github.com/jenkinsci/bom/pull/393/checks?check_run_id=1643221348 (CC @timja)

Reply:
https://github.com/jenkinsci/git-plugin/pull/1023 applies a fix for that case.  I'll try to release a new version of the git plugin with that fix within the next 7 days.

Reply:
Git plugin 4.5.2 released with this fix to the test.